Replicator/Worker.hh

```diff
@@ -134,9 +134,10 @@ namespace litecore { namespace repl {
 
         // Add the token for collection info to the format string
         static inline const char* formatWithCollection(const char* fmt) {
-            const std::string fmtStr = format("%s %s", kCollectionLogFormat, fmt);
-            const auto        found  = _formatCache.insert({fmtStr, 0});
-            return found.first->first.data();
+            std::unique_lock<std::mutex> lock(_formatMutex);
```

Comment:
It gets more and more expensive now. Consider use read/write lock. I assume most read would be successful.

Reply:
What do you mean exactly? I thought of doing:
- check for existing string
- if it exists already, return pointer to it
- if it doesn't, lock mutex and insert the string, then return a pointer to it

The issue with this, is if we are hitting the case which is not thread-safe, those two threads are trying to insert the same value at the exact same time. This means, it's pretty likely that the log string might not exist yet, both threads see that it doesn't exist as they're checking at the same time, and both threads will again try to insert at the same time.

I also would not worry about performance. I just did some performance testing, and over an average of 5,000 calls, that function still only takes ~1.5 microseconds - that is including the mutex lock.

Reply:
Oh, your preceding 3 bullets smell something fishing: it's not thread safe to read outside lock. The first bullet ought to be, check for existing string inside readLock.
By the way, from 0.9 to 1.5 is a 30% increase.

Reply:
Are you saying that STL container may handle one-write and multi-read concurrently ?

Reply:
Yes. Additionally, a scenario which you may think is dangerous:

thread A wants to log "Finished!"
thread B also wants to log "Finished!"

both threads read concurrently, "Finished!" does not exist. `mutex` is always thread-safe, so only one of them (say A) can grab it. A calls `_formatCache.insert()` and inserts "Finished!" to the cache. Now A releases the mutex, and B grabs it, then calls `_formatCache.insert()`. As `_formatCache.insert()` does a lookup first, it finds that "Finished!" already exists, so it does no insertion and just returns the iterator to the existing element.

Reply:
From cppreference:

> All const member functions can be called concurrently by different threads on the same container. In addition, the member functions begin(), end(), rbegin(), rend(), front(), back(), data(), find(), lower_bound(), upper_bound(), equal_range(), at(), and, except in associative containers, operator[], behave as const for the purposes of thread safety (that is, they can also be called concurrently by different threads on the same container)

And:

> Different elements in the same container can be modified concurrently by different threads, except for the elements of std::vector<bool> (for example, a vector of [std::future](https://en.cppreference.com/w/cpp/thread/future) objects can be receiving values from multiple threads).

Reply:
My understanding is that STL won't touch the elements themselves. So, different elements, provided you already have the references to them, can be modified concurrently. But STL does not protect the wiring from concurrent read/write.
"find" will go through the wiring of the container, "insert", may modify the wiring. My understanding is they cannot be performed concurrently.
But you may do "find" inside a readLock, and only it fails do a "insert" under normal lock.

Reply:
unordered_set is based on a hash-map. Elements are placed in "buckets" based on the hash of the element. The container is not consecutive or ordered, so insertion of any element does not invalidate held iterators. There is one scenario in which iterators can be invalidated, described here:
> If rehashing occurs due to the insertion, all iterators are invalidated. Otherwise iterators are not affected. References are not invalidated. Rehashing occurs only if the new number of elements is greater than max_load_factor() * bucket_count()

Additionally:
>References and pointers to data stored in the container are only invalidated by erasing that element, even when the corresponding iterator is invalidated.

So the pointers returned from `formatWithCollection()` are never invalidated.

Reply:
If we are worried about unknown behaviour of the following scenario:
- One thread `find`s the string it's looking for, has the `found` iterator which points to it.
- Before it can return `found->data()`, another thread performs an `insert()` operation which exceeds the available capacity of the `unordered_set`, and a rehash occurs - thus invalidating the reading thread's iterator before it got chance to return.
We can use an `std::shared_mutex` to allow multi-read and single-write locking.

Reply:
I am not worried about the pointer returned by "find". What I am concerned is, during the operation of "find", it may read the container's wiring, which "insert" may modify. In this case, it's the hash table itself.

Reply:
That's what I mean. One thread may have an iterator returned from `find`, which may be invalidated if another thread performs an `insert` that causes a rehash of the container. I am currently modifying the function to use a shared_mutex which allows multiple reads, but blocks reads when a write needs to occur.

Reply:
(:- The last 2nd and 3rd msgs are example of concurrent posts, orders got screwed.